---
title: "Intro to Seurat Workshop"
author: "Nguyen, Quang N"
date: "Last compiled on `r format(Sys.time(), '%d %B %Y')`"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
    number_sections: yes
    highlight: haddock
    df_print: paged
editor_options:
  chunk_output_type: inline
---

# REFERENCES

https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#cluster-the-cells-1


# INSTALL PACKAGES 

```{r setup, include=FALSE, echo=FALSE, results='hide'}

#knitr::opts_chunk$set(echo = FALSE, collapse=TRUE, out.width = "70%",out.height = "70%", fig.align = "center")

#install.packages(c('Seurat','dplyr','patchwork','Matrix', 'pastecs', 'ggplot2','clustree'))

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install("glmGamPoi")

devtools::install_github("satijalab/sctransform", ref = "develop")

library(Seurat)
BiocManager::install('limma')

library(clustree)

library(dplyr)
library(Matrix)
library(pastecs)

library(ggplot2)
library(patchwork)

```

# LOAD INPUT DATA

Input data = Non-normalized UMI count matrix from RNA-seq of 2,700 single PBMCs from 10X cellranger pipeline

Please direct to "/filtered_gene_bc_matrices/hg19" from your download of https://github.com/denvercal1234GitHub/learntoRfromNothing

```{r}
# Load the PBMC dataset, containing matrix.mtx; barcodes.tsv; genes.tsv
pbmc.data <- Read10X(data.dir = "/Users/clusteredatom/Documents/YouTube_Video/LearntoRfromNothing/2022Nov11_IntroToSeuratWorkshop_QNN/filtered_gene_bc_matrices/hg19")

dim(pbmc.data)
```

***Granulysin:*** Cytolytic and proinflammatory molecule first identified by a screen for genes expressed ‘late’ (3–5 days) after activation of human peripheral blood mononuclear cells. Granulysin is present in cytolytic granules of cytotoxic T lymphocytes and natural killer cells (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2679253/)

***MS4A1 (CD20):*** This gene encodes a B-lymphocyte surface molecule which plays a role in the development and differentiation of B-cells into plasma cells (https://www.proteinatlas.org/ENSG00000156738-MS4A1)

```{r}
# Quick check if your genes of interest are detected at all

interesting_genes <- c("CD3E","CD4", "CD8A","MS4A1","GNLY") 

interesting_genes %in% rownames(pbmc.data)
```


# FILTER GENES (min.cells)

***Expect rare cell types?*** Be permissive and retain genes that are expressed in few cells

***Technical notes:***
counts>0 returns a matrix where each entry is TRUE/FALSE if that entry of the counts matrix exceeds 0. 
Performing rowMeans on that matrix returns the number of cells with a count > 0 divided by total # of cells for each gene, which is the percent of cells expressing a gene. 
Performing rowSums on that matrix returns the total number of cells having a detectable expression for that gene

```{r}
#3 is the number of cells at or below which 60% of the genes got detected
quantile((Matrix::rowSums(pbmc.data>0)), probs = c(0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9))
```

```{r}
#log10(quantile((Matrix::rowSums(pbmc.data>0)), probs = 0.6))
#When logging count data, the + 1 is used to avoid log10(0) which is not defined
hist(log10((Matrix::rowSums(pbmc.data>0))+1), main=expression("Many genes epxressed in less than n = log"^"-1"*""[10]*"(0.48) ~ 3 cells"), col='wheat', labels=FALSE, ylab = "Frequency of genes", xlab = "Number of cells with positive expression for every gene")  
abline(v=log10(quantile((Matrix::rowSums(pbmc.data>0)), probs = 0.6)), col = "red", lty=2, lwd = 2)

```


## Remove low-quality genes

min.cells: Include features detected in at least this many cells.

min.features: Include cells where at least this many features are detected.

names.delim: For the initial identity class for each cell, choose this delimiter from the cell's column name. E.g. If your cells are named as BARCODE-CLUSTER-CELLTYPE, set this to “-” to separate the cell name into its component parts for picking the relevant field.

```{r}
# Initialize the Seurat object with the raw (non-normalized data)
# Original analysis set min.features = 200
pbmc <- CreateSeuratObject(counts = pbmc.data, names.delim = "-", names.field = 2, min.cells = quantile((Matrix::rowSums(pbmc.data>0)), probs = 0.60))

pbmc
```


```{r}
dim(pbmc.data)
```


# FILTER CELLS

## nFeatures threshold

The number of unique genes detected in each cell (nFeatures).
1. Too low nFeatures = Low-quality cells or empty droplets 
2. Too high nFeatures = Cell doublets or multiplets 

***Total number of molecules detected within a cell (nCounts) correlates strongly with nFeatures***

min.features parameter or setting threshold for nFeatures (library complexity) helps filter out cells likely have random barcodes encapsulated without any cell present. Usually, cells with less than 100 genes detected are filtered out.

This threshold varies with cell type, e.g., higher threshold is appropriate for brain cells than PBMCs that can go up to 600 genes/cell!

For high quality data, the proportional histogram should contain a single large peak that represents cells that were encapsulated. 
***If we see a small shoulder to the left of the major peak, or a bimodal distribution of the cells:***
1. Cells that failed for some reason. 
2. There are biologically different types of cells (i.e. quiescent cell populations, less complex cells of interest), and/or one type is much smaller in size than the other (i.e. cells with high counts may be cells that are larger in size). ***Therefore, this threshold should be assessed with other metrics that we describe in this lesson.***

```{r}
# Count the gene as 1 only if it has non-zero reads mapped (this is not summing the values)
numberofgenes_per_cell <- Matrix::colSums(pbmc.data>0) 

#summary(numberofgenes_per_cell)

# When logging count data, the + 1 is used to avoid log10(0) which is not defined
hist(log10(numberofgenes_per_cell+1), main='Log number of expressed genes per cell', col='wheat')
```


```{r}
#nFeature
#quantile(pbmc@meta.data[,3], na.rm=TRUE, probs = c(0.000001, 0.05, 0.1, 0.15, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.85, 0.9, 0.95, 0.96, 0.97, 0.98, 0.99))

summary(pbmc@meta.data[,3])
```


```{r}
?VlnPlot
VlnPlot(pbmc, features = "nFeature_RNA", cols = "green") + 
  scale_y_log10 (breaks= seq(0, 2500, by = 500)) + 
  geom_hline(yintercept=c(2500, 200), linetype="dashed", color = "blue") + labs(caption = "Thresholds (200, 2500)")
```

## percent.mt threshold

The percentage of all the counts (i.e., counts or transcripts) that map to the mitochondrial gene group for each cell.

Column sum of the matrix in the counts slot for features belonging to the gene group, divided by the column sum for all features, multiplied by 100.

High mitochondrial contamination = Potentially dead or dying cells, but they might represent distinct cell state/types (e.g., in respiratory processes).

In a majority of cases, it is sufficient to filter out cells with percent.mt > 20 or so, depending on your data (e.g, higher threshold for single-nucleus RNA-seq data).

***After filtering, you can regress out percent.mt during normalization if strong effects visualized on your UMAP or PCA plots.*** 

```{r}
pbmc <- PercentageFeatureSet(pbmc, pattern = "^MT-", col.name="percent.mt")
```


```{r}
#stat.desc(pbmc@meta.data)
#colnames(pbmc@meta.data)

#percent.mt
quantile(pbmc@meta.data[,4], na.rm=TRUE, probs = c(0.05, 0.1, 0.15, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.85, 0.9, 0.95, 0.96, 0.97, 0.98, 0.99))
```


```{r}
# Visualize the distribution of mitochondrial gene expression detected per cell
#aes(color=sample, x=mitoRatio, fill=sample)
pbmc@meta.data %>%
  	ggplot(aes(x=percent.mt)) + 
  	geom_density(alpha = 0.2)  + 
  	theme_classic() +
  	geom_vline(xintercept = 5)
```

## Visualize combined QC metrics

```{r}
FeatureScatter(pbmc, feature1 = "percent.mt", feature2 = "nFeature_RNA") + 
    geom_vline(xintercept = 5) +
  	geom_hline(yintercept = c(200, 2500)) +
    labs(title="Very high percent.mt cells\n have very low nFeatures") + NoLegend() -> p1

FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") +
  	geom_hline(yintercept = c(200, 2500)) +
    labs(title="nFeature and nCount are correlated") + NoLegend() -> p2

p1+p2
```


## Remove low-quality cells

```{r}
#Check which cells we eliminated by our combined QC thresholds
length(WhichCells(pbmc, expression = nFeature_RNA <= 200 | nFeature_RNA >= 2500 | percent.mt > 5))
```

```{r}
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)

```


```{r}
save(pbmc.data, file="/Users/clusteredatom/Documents/YouTube_Video/LearntoRfromNothing/2022Nov11_IntroToSeuratWorkshop_QNN/pbmc.data.RData")

save(pbmc, file="/Users/clusteredatom/Documents/YouTube_Video/LearntoRfromNothing/2022Nov11_IntroToSeuratWorkshop_QNN/pbmc_preNormalize.RData")
```


# SUMMARIZE SELECTED CELLS AND GENES

```{r}
pbmc$groups <- sample(c("group1", "group2", "group3"), size = ncol(pbmc), replace = TRUE)

head(pbmc@meta.data)
```


```{r}
#%>% replace(is.na(.), 0) %>% mutate(row_sum = rowSums(across(where(is.numeric))))
#%>% summarise_all(funs(sum))
#https://dplyr.tidyverse.org/articles/colwise.html
  pbmc@meta.data %>% group_by(groups) %>% summarise(num_selected_cells = n(), 
             mean_library_size = mean(nCount_RNA),
             mean_num_detected_genes = mean(nFeature_RNA),
             mean_mitochondrial_fraction = mean(percent.mt)) %>% janitor::adorn_totals()
```


```{r}
interesting_genes %in% rownames(pbmc)
```


```{r}
#Check how many cells having positive expression for a gene
length(WhichCells(pbmc, expression = `BCL6` > 0))
```


# OTHER QC METRICS

***Sequencing depth:***
If using read counts rather than UMI counts, Matrix::colSums(cxcr5poscd8.data) would be the total number of reads (sequencing depth).
Wells with few reads/molecules are likely to have been broken or failed to capture a cell.

***Novelty (complexity) score:*** 
Ratio of the log10 of number of genes per cell (nFeatures) over the log10 of number of UMIs per cell (nCount). 

Low complexity cells could represent a  cell type (e.g., red blood cells which lack a typical transcriptome), an artifact, or contamination. Generally, we expect the novelty score to be above 0.80 for good quality cells.

If there are many captured transcripts (high nUMI) and a low number of genes detected in a cell, could mean we only captured a low number of genes and simply sequenced transcripts from those lower number of genes over and over again. 

***Percent of ribosomal gene expression***

***Predicted doublets***

***Cell sampling differences***

```{r}
pbmc@meta.data %>% 
  	ggplot(aes(x=groups, fill=groups)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("Number of cells")
```

# AUTOMATED QC METHOD

miQC is data-driven QC metric that jointly models both the proportion of reads mapping to mtDNA and the number of detected genes with mixture models in a probabilistic framework to predict which cells are low-quality in a given dataset

Please see: https://github.com/greenelab/miQC

# REGRESS OUT UNWANTED VARIATIONS DURING NORMALIZATION

## Theory

### Why do we log-transform data?
Log-transformation can:
1. Reduce the variance on the first PC (i.e., variance stabilization; to visualize batch effect based on counts) and already separates some biological effects. 
2. Makes the distribution of the expression values more normal.

BUT, log-transformation might not be sufficient to account for different technical factors between the cells (e.g., sequencing depth). 
Therefore, need to also normalise also by library size.

### LogNormalize() method
In seurat, global-scaling normalization method ***(LogNormalize)*** normalizes the expression of a gene for each cell by the total expression (colSums) to account for sequencing depth.
Then, multiplies by a scale factor (10,000 as if all cells have 10k UMIs overall), and log-transforms the result.

ScaleData converts normalized gene expression to Z-score (values centered at 0 and with variance across cells of 1). As a result, it gives equal weight in downstream analyses, so that highly-expressed genes do not dominate.

### SCTransform() method

***SCTransform() method simultaneously performs variance stabilization and regresses out unwanted variation by normalizing and scaling the gene expression values to account for differences in sequencing depth and overdispersed count values (nCounts) while setting the "vars.to.regress" (using linear model to perform a second regression and regresses out the effect of a given covariate)*** 

The goal of SCTransform is to perform normalization within an experiment by learning the model of technical noise (within the experiment). It still uses regularized negative binomial (regularization being performed on the overdispersion parameter such that genes with similar mean expression have similar overdispersion).

SCTransform stabilizes the variance of the data by regressing out the effect of the sequencing depth from each cell. Specifically, it uses pearson residuals from “regularized negative binomial regression,” where cellular sequencing depth is utilized as a covariate in a generalized linear model. It is shown to successfully remove the influence of technical characteristics from downstream analyses while preserving biological heterogeneity.

SCTransform provide a more parsimonious regularization and improved the variable feature selection procedure. 

For corrected counts, with multiple datasets, SCTransform now performs correction based on minimum of the median UMI. This is useful if you wish to perform DE across two different SCT normalized datasets with differences in sequencing depth (See https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02584-9) 


***vars.to.regress in ScaleData() and SCTransform()***:
Specifies variables to regress out of the scaled data or sctransform residuals, preventing these variables from contributing much to the PCA. 

1. Typically, if you have removed cells with a high percentage of mitochondria, you don't need to regress out "percent.mt".***

2. If you have a set of genes, eg. the cell cycle genes, that you don't want to drive the clustering results, regress out those genes before running PCA, or remove those genes from the set of variable features used in the PCA.

By default, SCT only considers gene expressed if they are expressed (UMI >=1) in a minimum of 5 cells (min.cells=5). We fix the slope parameter of the GLM to ln(10) with log10(total UMI) used as the predictor. 

***Technical note of SCTransform:***
We utilize an improved parameter estimation procedure that alleviates uncertainty and bias that result from fitting GLM models for very lowly expressed genes.
We place a lower bound on gene-level standard deviation when calculating Pearson residuals. This prevents genes with extremely low expression (only 1-2 detected UMIs) from having a high pearson residual.


### latent.vars during DEG analysis

Setting latent.vars in FindMarkers() specifies latent variables to include in the differential expression model. 

The normalized data, rather than scaled data or sctransform residuals, is typically used for differential expression.

# CLUSTERING

## Normalize, dimentionality reduction, and construct KNN graph

Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’.

As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors() function, and takes as input the previously defined dimensionality of the dataset (a number of PCs).

The goal of non-linear dimensional reduction techniques, such as tSNE and UMAP, is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. 

```{r}
SCTransform(pbmc, vst.flavor = "v2", verbose = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE) %>%
    RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
    FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) -> pbmc_postNormalize

for(i in seq(0,1.5,0.5)){
  pbmc_postNormalize <- Seurat::FindClusters(pbmc_postNormalize, resolution = i)
}

save(pbmc_postNormalize, file="/Users/clusteredatom/Documents/YouTube_Video/LearntoRfromNothing/2022Nov11_IntroToSeuratWorkshop_QNN/pbmc_postNormalize.RData")

pbmc_postNormalize@meta.data
```

## Cluster cells 

To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM [SLM, Blondel et al., Journal of Statistical Mechanics], to iteratively group cells together, with the goal of optimizing the standard modularity function. 

The FindClusters() function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. 

***We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells.*** Optimal resolution often increases for larger datasets. 

### Select clustering resolution (clustree package)

Please see: https://github.com/lazappi/clustree

To build a clustering tree we need to look at how cells move as the clustering resolution is increased. 

Each cluster forms a node in the tree and edges are constructed by considering the cells in a cluster at a lower resolution (say k=2) that end up in a cluster at the next highest resolution (say k=3). 

By connecting clusters in this way we can see how clusters are related to each other, which are clearly distinct and which are unstable. 

Extra information about the cells in each node can also be overlaid in order to help make the decision about which resolution to use. 

For more information about clustering trees please refer to our associated publication (Zappia and Oshlack 2018).

### By cluster stability

The stability index from the {SC3} package (Kiselev et al. 2017) measures the stability of clusters across resolutions and is automatically calculated when a clustering tree is built. 

It can be accessed by setting an aesthetic to "sc3_stability" and because it is calculated by cluster we don’t need to provide an aggregation function. 

```{r}
clustree(pbmc_postNormalize, prefix = "SCT_snn_res.", 
         node_colour = "sc3_stability", use_core_edges = T, 
         node_label = NULL, node_label_aggr = NULL)   +
    scale_edge_color_continuous(low = "blue", high = "red") +
    guides(edge_colour = FALSE, edge_alpha = FALSE) +
    theme(legend.position = "bottom")
```

### By gene expression 

Use the expression of individual genes. 

```{r}
#colnames(pbmc_postNormalize@meta.data)

clustree(pbmc_postNormalize, prefix = "SCT_snn_res.", node_colour = "PPBP", node_colour_aggr = "median")
```

clustree(pbmc_postNormalize, prefix = "SCT_snn_res.", node_colour = "PPBP", node_colour_aggr = "median")


### FindCluster()

```{r}
Idents(pbmc_postNormalize) <- pbmc_postNormalize$SCT_snn_res.0.5

table(pbmc_postNormalize@active.ident, pbmc_postNormalize$SCT_snn_res.0.5)
```
# QC CLUSTERS

Identify potential confounding variable, besides your experimental design


```{r}
DimPlot(pbmc_postNormalize, reduction = "umap", label = T, label.size = 5, repel = F, pt.size = 0.5)
```


```{r}
DimPlot(pbmc_postNormalize,
        reduction = "umap",
        group.by = "groups") +
    labs(title="UMAP by groups")

```


```{r}
FeaturePlot(pbmc_postNormalize, features = "percent.mt") + theme(plot.title = element_text(size=10)) 
```

# CLUSTER ANNOTATION 

```{r}
plot <- DimPlot(pbmc_postNormalize, reduction = "umap")

cluster4_Cells <- CellSelector(plot = plot)

```
FindMarkers output a data.frame with a ranked list of putative markers as rows, and associated statistics as columns (p-values, ROC score, etc., depending on the test used (test.use)). The following columns are always present:

avg_logFC: log fold-chage of the average expression between the two groups. Positive values indicate that the gene is more highly expressed in the first group

pct.1: The percentage of cells where the gene is detected in the first group

pct.2: The percentage of cells where the gene is detected in the second group

p_val_adj: Adjusted p-value, based on bonferroni correction using all genes in the dataset

```{r}
#Idents(pbmc_postNormalize, cells = cluster4_Cells) <- "NewCells"

?FindMarkers
cluster4_DEGs <- FindMarkers(pbmc_postNormalize, ident.1 = "4")

cluster4_DEGs

```

```{r}
VlnPlot(pbmc_postNormalize, features = interesting_genes)
```

```{r}
FeaturePlot(pbmc_postNormalize, features =  c( "MS4A1", "CD79B"), blend=T) 
```


